readme.md

# Neo Tank Trouble v0.10
A wip reimagining of Tank Trouble. 


## Controls

Use the arrow keys to navigate the menu, and ctrl to select.

- In the lobby, shoot to join.

	Red tank uses the arrow keys and ctrl to shoot.
	Blue tank uses ESDF and Q to shoot.
	Green tank uses IJKL and Y to shoot.
	Yellow tank uses the number pad's 8456 and 0 to shoot.
	Press Shift or Escape to pause, and Shift or Escape to back out of a submenu / unpause.

Collect rainbow powerups, shoot the other tanks, and be that last one standing to win. 
Shooting walls will damage them until they disappear. 
The pink circle will shrink throughout the round, killing any tank in it's path.


## Running

Download the source. Opening `index.html` is enough to run the game. 





changelog.md
# Changelog

- v0.10 Jan 12 2023 --
	- Tweaked Yellow's color
	- Lowed turn acceleration

- v0.9, Jan 9 2023 --
	- Added an attempt at showing controls in the lobby
	- Added an option to go back to the lobby in the menu
	- Removed the "Round Positions" option, it's now on permanently(for walls)

- v0.8, Jan 6 2023 --
	- Happy (late) new year!
	- I rewrote the entire game, and wrote a custom game engine to go with it, meaning the game is officially disconnected from p5.js! That means quite a few things, namely:
		- It'll be easier to put this game onto a 'proper' website
		- I get full control over the performance of various things, and any engine level optimizations I'm able to implement, I'll be able to do(the game is already notably faster than with the p5js dependancy)
		- It should also be a bit faster, except for the menu which got slower. I really tried my best to speed up the menu, but it just isn't fast and I'm unsure why.
		- A *lot* changed, and I might miss some here in this list
	- A lobby of sorts was added. Press the shoot button of any tank for them to join. 
	- When you die, a little marking of your color will be left, your corpse
	- Tank graphics were tweaked very slighty, and the colors were changed to be more saturated
	- The maze does a nice animation between rounds, instead of instantly switching the maze
	- Specifically, the maze will 'animate' the maze generation(it's completely fake), and players corpses will be smoothly moved to their new position, and any storms will fade out
	- Graphics should (hopefully) look more crisp
	- The storm will slide in
	- Ministorm bullets have an animation now
	- Ministorms will bounce outward when created
	- Increased storm start radius
	- Tank acceleration speed up a bit
	- Walls are a bit brighter
	- The laser's collision is a more accurate and has a maximum range now
	- The laser shot graphic was improved slightly
	- Turret buffed a lot, it shoots faster and reacts quicker
	- Turret graphic changed slightly
	- Increased the spawn rate of power ups slightly
	- Switched the default maze algorithm to Cellular x2

- v0.7, Dec 28 2022 --
	- Added slight acceleration to turning and movement
	- Increased tank turning speed
	- Walls are slightly brighter now

- v0.6, Dec 27 2022 --
	- Updated menu sliders to not scale it's text value
	- Renamed "Power Up Chance" option to "Power Up Amount", and moved it to the "Customize Power Ups" sub menu.
	- The machine gun, turret, and ministorm powerups are slightly rarer.
	- You can now customize the chances of any powerup spawning in(Note: Setting all the chances to zero will make all powerups machine guns. To turn off powerups, set "Power Up Amount" to "Off")
	- Pause key is also mapped to Shift
	- When the camera is zoomed out enough, a graphic will appear in front of each tank showing what direction their facing. You may turn this off in settings.
	- Small bullet's graphics are clamped more to make them more visible when the camera is zoomed out
	- Attempted to make the camera less weird, it will now take the average of everyone's position + the point between the 2 furthest apart, instead of only the 2 furthest apart from each other. I expect there to be issues with players being able to escape the camera, but I haven't found any yet aside from extreme cases. I'll continue adjusting.
	- Added new map generation options, you can now choose between 3 maze generation algorithms -
		- Prim's algorithm, the original default one. It leads to many twists and turns, and the occasional straight path. The wall breaking mechanics are well suited to this one.
		- Cellular Automota, which can generally generates more 'chaotically' than Prim, often generating straight paths and the occasional small room. You can configure it to generate mostly straight paths as well. You may also want to mess araound with it's iterations, which will allow it more time to generate, sometimes leading to more straighter paths, though it may increase map generation times. Setting it's iterations to 0 is identical to using the Random algorithm set to 5.
		- Random, which is just uniformly random. You can set the general density of it, but you can't set it to anything higher than 8, since that may lead to issues when trying to find a place to put any player.
		- I eventually want to add more algorithms, so. Yea. Gamaing
	- Fixed issue where turrets just were not shooting anything
	- Slowed down tank turning speed(both normal and laser focus) slightly to make aiming a bit easier
	- The bomb's timer is shorter
	- The turret now waits 1 second before firing after being createed, down from the previous ~4 seconds

- v0.5, Dec 23 2022 --
	- Added a "Show FPS" setting, under a new graphics section
	- Walls are now rounded to integer positions, which speeds up rendering quite a bit. You can turn it off in settings if it's too disorientating. You can also choose to round all(most) positions, which very likely doesn't do anything for performance.
	- Walls are rendered differently and less dumbly. I'm able to get 60 fps with a size 64 map fully zoomed out on my terrible laptop now.
	- Walls are a bit brighter, particularly at one health left, to make them easier to see
	- Fixed laser insta-killing you if you use it into a wall, instead it's timer is shortened. I'm sure this won't be abused :)
	- Lasers won't be instantly destroyed if it hits someone
	- Added Turret powerup

- v0.4, Dec 22 2022 --
	- Camera will now consider tank direction
	- Camera scale is smaller and capped
	- Added Shotgun powerup
	- Added Mine powerup
	- Bomb's wall damage radius made much bigger, player kill radius unchanged
	- Shrank the storm powerup's graphic
	- Fixed far bottom and right walls not being destroyed by bombs
	- Clamped powerups will size themselves based on how far they are
	- Reorganized files

- v0.3, Dec 19 2022 --
	- Eased border flashing
	- Added Machine Gun powerup
	- Added Laser powerup
	- Added Bomb powerup

scripts/ui.js
class UI extends Entity {
  constructor(_) {
    super(_);
    
    this.menuList = [];
    
    this.menu = new Menu();
    
    // if (ismain) this.menu
    //   .add_label('Play with:')
    //   .add_button('2 players', () => {
    //     startMatch(2);
    //   })
    //   .add_button('3 players', () => {
    //     startMatch(3);
    //   })
    //   .add_button('4 players', () => {
    //     startMatch(4);
    //   })
    // else 
      this.menu
      .add_button('Continue', () => {
        entity_destroy(this);
      })
    if (!game.lobby)
      this.menu
      .add_button('Kill All', () => {
        entity_destroy(Player)
        entity_destroy(this)
      })
      .add_button('Lobby', () => {
        game.doLobby = true;
        entity_destroy(Player);
        entity_destroy(this)
      })
    
    this.menu
      .add_label('Game Settings:')
      .add_slider('Map Size', 8, 64, 4, settings.mapSize, undefined, v => {
        settings.mapSize = v
      })
      .add_button("Customize Maze Generation", () => {
        this.open(this.mazeMenu)
      })
      .add_slider('Storm Speed', 0, 20, 1, settings.stormSpeed, undefined, v => {
        settings.stormSpeed = v
      })
      .add_button("Customize Power Ups", () => {
        this.open(this.powerUpMenu)
      })
      .add_label('Accessibility:')
      .add_radio('Flashing', ['No', 'Yep'], settings.flashing, v => {
        settings.flashing = v;
      })
      .add_radio('Screen Shake', ['Off', '50%', '100%'], settings.screenShake, v => {
        settings.screenShake = v;
      })
      .add_radio('Contrast', ['+0', '+1', '+2'], settings.contrast, v => {
        settings.contrast = v;
      })
      .add_label('Graphics:')
      .add_radio('Show Tank Direction', ['Don\'t', 'Yes'], settings.showTankDir, v => {
        settings.showTankDir = v;
      })
      .add_radio('Show FPS', ['Nope', 'Heck yea'], settings.showFPS, v => {
        settings.showFPS = v;
      })
      // .add_button('Save Settings', () => {
      //   console.log(JSON.stringify(settings))
      //   window.localStorage.setItem('settings', JSON.stringify(settings))
      // })
      // .add_button('Clear Save', () => {
      //   window.localStorage.removeItem('settings');
      // })
      //.add_slider('Test', 0, 10, 1, 0)
    //this.menu.position = 1;
    
    this.mazeMenu = new Menu();
    this.mazeMenu
      .add_button('Back', () => this.pop())
      .add_label('Maze Customization')
      .add_radio('Maze Algorithm', ['Prim', 'Cellular', 'Random'], settings.mapAlgorithm, v => settings.mapAlgorithm = v)
      .add_radio('Cellular - Chaos', ['0.5x', '1x', '2x'], settings.mapCellularStraight, v => settings.mapCellularStraight = v)
      .add_slider('Cellular - Iterations', 0, 200, 5, settings.mapCellularIterations, undefined, v => settings.mapCellularIterations = v)
      .add_slider('Random - Density', 0, 8, 1, settings.mapRandomChance, undefined, v => settings.mapRandomChance = v)
    
    this.powerUpMenu = new Menu();
    this.powerUpMenu
      .add_button('Back', () => {
        this.pop()
      })
      .add_label('Power Up Customization')
      .add_radio('Power Up Amount', ['Off', '0.5x', '1x', '2x'], settings.powerUps, v => {
        settings.powerUps = v;
      })
      .add_slider('Machine Gun Chance', 0, 10, 1, settings.powerUpWeights[0], undefined, v => settings.powerUpWeights[0] = v)
      .add_slider('Mine Chance', 0, 10, 1, settings.powerUpWeights[1], undefined, v => settings.powerUpWeights[1] = v)
      .add_slider('Laser Chance', 0, 10, 1, settings.powerUpWeights[2], undefined, v => settings.powerUpWeights[2] = v)
      .add_slider('Turret Chance', 0, 10, 1, settings.powerUpWeights[3], undefined, v => settings.powerUpWeights[3] = v)
      .add_slider('Shotgun Chance', 0, 10, 1, settings.powerUpWeights[4], undefined, v => settings.powerUpWeights[4] = v)
      .add_slider('Bomb Chance', 0, 10, 1, settings.powerUpWeights[5], undefined, v => settings.powerUpWeights[5] = v)
      .add_slider('MiniStorm Chance', 0, 10, 1, settings.powerUpWeights[6], undefined, v => settings.powerUpWeights[6] = v)
    this.powerUpMenu.position = 0;
    
    this.ignorePause = true;
    this.depth = 1000
    this.scroll = new Sod(4, 0.6, 0);
    this.imlazy = false;
    
    this.open(this.menu);
  }
  open(menu) {
    // bad idea
    let fix = 20;
    while (fix-- > 0 && menu.elements[menu.position] instanceof MenuLabel)
      menu.position++;
    menu.scrollY = new Sod(2, 1, 0);
    this.menuList.push(menu);
  }
  pop() {
    this.menuList.pop();
  }
  step() {
    const currentMenu = this.menuList[this.menuList.length - 1];
    currentMenu.scroll(
      controls[0].check_stutter('down', 14, 6) - 
      controls[0].check_stutter('up', 14, 6) );
    currentMenu.change(
      controls[0].check_stutter('right', 8, 2) - 
      controls[0].check_stutter('left', 8, 2) );
    if (controls[0].check_pressed('shoot'))
      currentMenu.click()
    
    this.scroll.update(1/60, (this.menuList.length - 1) * 256)
    
    setPause(2);
    if (this.imlazy)
      if (controls[0].check_pressed('pause'))
        this.pop()
    this.imlazy = true;
    if (this.menuList.length == 0)
      entity_destroy(this);
    
  }
  postdrawgui() {
    draw_push();
    if (this.menuList.length == 1 && !this.ismain) {
      draw_set_fill(util_toCssGrey(0, 150+ settings.contrast * 40));
      draw_nostroke()
      draw_rectangle(0, 0, surface_default.width, surface_default.height);
    }
    for (let i = 0; i < this.menuList.length; i++) {
      if (this.menuList.length > 1 && i == this.menuList.length-1) {
        draw_set_fill(util_toCssGrey(0, 150+ settings.contrast * 40));
        draw_nostroke()
        draw_rectangle(0, 0, surface_default.width, surface_default.height);
      }
      const currentMenu = this.menuList[i];
      currentMenu.scrollY.update(1/60, currentMenu.position - 1);
      currentMenu.draw(64 + i * (256) - this.scroll.value, 64 - currentMenu.scrollY.value * 12, 38);
    }
    //fill(30)
    //noStroke()
    //rect(0, height - 100, width, 100)
    draw_pop();
  }
}